sw/cheri/common/sdcard-utils.hh

```diff
@@ -0,0 +1,472 @@
+ */
+
+class SdCard {
+ private:
+  // Access to SPI controller.
+  volatile SonataSpi *spi;
+  // Access to GPIO block (required for SD card detection).
+  volatile SonataGpioBase *gpio;
+  // Chip select (single bit set).
+  uint32_t cs;
+  // SD card `detect` pin (single bit set).
+  uint32_t det;
+  // CRC_ON mode used?
+  // - this is off by default for SPI mode, but may be enabled for error detection in both the
+  //   host and the card.
+  bool crcOn;
+  // Access to diagnostic logging.
+  Log *log;
+
+  // We need to clock the device repeatedly at startup; this test pattern is used to ensure that
+  // we keep the COPI line high and it cannot be misinterpreted as a command.
+  static constexpr uint8_t ones[] = {0xffu, 0xffu, 0xffu, 0xffu, 0xffu, 0xffu, 0xffu, 0xffu, 0xffu, 0xffu};
+
+ public:
+  // Transfers in SPI mode are always in terms of 512-byte blocks.
+  static constexpr unsigned kBlockLen = 512u;
+
+  // Supplied with SPI controller and its appropriate CS line, and a GPIO block with its input pin
+  // number. The SPI controller performs the communication with the SD card and the GPIO simply
+  // provides the card detection indicator (low = present).
+  //
+  // CRC calculation is turned on by default, although this requires the CPU to perform a little
+  // bit of additional work. It may be disabled if desired. SPI mode communication may be performed
+  // with or without CRC checking.
+  //
+  // Logging may optionally be requested.
+  SdCard(volatile SonataSpi *spi_, volatile SonataGpioBase *gpio_, unsigned cs_ = 1u, unsigned det_ = 16u,
+         bool crc_ = true, Log *log_ = nullptr)
+      : spi(spi_), gpio(gpio_), cs(1u << cs_), det(1u << det_), crcOn(crc_), log(log_) {}
+
+  // SD command codes. (Section 7.3.1)
+  enum CmdCode {
+    CMD_GO_IDLE_STATE        = 0,
+    CMD_SEND_OP_COND         = 1,
+    CMD_SEND_IF_COND         = 8,
+    CMD_SEND_CSD             = 9,
+    CMD_SEND_CID             = 10,
+    CMD_STOP_TRANSMISSION    = 12,
+    CMD_SET_BLOCKLEN         = 16,
+    CMD_READ_SINGLE_BLOCK    = 17,
+    CMD_READ_MULTIPLE_BLOCK  = 18,
+    CMD_WRITE_SINGLE_BLOCK   = 24,
+    CMD_WRITE_MULTIPLE_BLOCK = 25,
+    SD_SEND_OP_COND          = 41,
+    CMD_APP_CMD              = 55,
+    CMD_READ_OCR             = 58,
+    CMD_CRC_ON_OFF           = 59,
+  };
+
+  // SD Control Tokens. (Section 7.3.3)
+  enum CtrlToken {
+    // Start Block Token precedes data block, for all but Multiple Block Write.
+    StartBlockToken = 0xfeu,
+    // Start Block Token used for Multiple Block Write operations.
+    StartBlockTokenMW = 0xfcu,
+    // Stop Transaction Token, for Multiple Block Writes.
+    StopTranToken = 0xfdu,
+  };
+
+  // Indicates whether there is an SD card present in the slot.
+  bool present() const { return !(gpio->debouncedInput & det); }
+
+  void select_card(bool enable) { spi->cs = enable ? (spi->cs & ~cs) : (spi->cs | cs); }
+
+  // Initialise the SD card ready for use.
+  bool init() {
```

Comment:
In my experience working with SD cards, it's better to actually do the initialization properly because you will find cards in the while that enforce the spec to the letter.

Reply:
Thanks for all the feedback; the code was originally a quick prototype just to check that we could use the SD card at all before 1.0, and it was based upon rather ad hoc tutorials/code found online, without having a proper specification.

It's only example code, but perhaps - at our leisure - we should shore it up properly; @GregAC plans to improve the robustness and error detection of the code.